### ISSUE TITLE
mrview - sometimes segfault upon loading data, sometimes not  on ubuntu 16.04

### ISSUE THREAD

**Issue by samuelstjean** [MRtrix3/mrtrix3#793]
Hello,

I think this one is likely related to bad libraries/config stuff, but I mostly want to see if anyone else has the same issue. Basically, whenever I try to load an image into mrview, sometime it works, sometimes I get a segfault for exactly the same image, 2-3 times in a row. It actually used to work before on the same computer which was running ubuntu 14.04, and I can't get it to work reliably on 16.04.

I of course pulled the latest master this morning, did a build clean, configure, build again. I can start mrview, but not view stuff in it. As it used to work before, I suspect bad lib/compiler combination, here is what I got

~~~bash
samuel ~/git/mrtrix3 $ [mrtrix3 (master)] ./configure 

MRtrix build type requested: release

Detecting OS: linux
Checking for C++11 compliant compiler [g++]: 5.4.0 - tested ok
Detecting pointer size: 64 bit
Detecting byte order: little-endian
Checking for variable-length array support: yes
Checking for non-POD variable-length array support: yes
Checking for zlib compression library: 1.2.8
Checking for Eigen 3 library: 3.2.92
Checking shared library generation: yes
Checking for Qt moc: moc (version 5.5.1)
Checking for Qt qmake: qmake (version 5.5.1)
Checking for Qt rcc: rcc (version 5.5.1)
Checking for Qt: 5.5.1
writing configuration to file './release/config': ok
~~~

And for the weird part, it just seems to work sometimes and sometimes not. This is one that segfaults.

~~~bash
samuel my_secret_folder_path_which_contains_too_much_information $ [DICOM_gab] /home/samuel/git/mrtrix3/release/bin/mrview 2.nii.gz -debug
mrview: [DEBUG] No config file found at "/etc/mrtrix.conf"
mrview: [DEBUG] No config file found at ".mrtrix.conf"
mrview: [INFO] GL renderer:  Quadro K2200/PCIe/SSE2
mrview: [INFO] GL version:   3.3.0 NVIDIA 367.48
mrview: [INFO] GL vendor:    NVIDIA Corporation
mrview: [DEBUG] loading font into OpenGL texture...
mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location = 0) in vec2 pos;
layout(location = 1) in vec2 font_pos;
uniform float scale_x;
uniform float scale_y;
out vec2 tex_coord;
void main () {
  gl_Position = vec4 (pos[0]*scale_x-1.0, pos[1]*scale_y-1.0, 0.0, 1.0);
  tex_coord = font_pos;
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
in vec2 tex_coord;
uniform sampler2D sampler;
uniform float red, green, blue;
out vec4 color;
void main () {
  color.ra = texture (sampler, tex_coord).rg;
  color.rgb = color.r * vec3 (red, green, blue);
}

mrview: [DEBUG] font loaded
mrview: [INFO] opening image "2.nii.gz"...
mrview: [DEBUG] sanitising image information...
Segmentation fault (core dumped)
~~~

It seems to complain about the image, but this is a freshly converted dicom with mrconvert though. It does this for a bunch of other images also (from scanner dicoms, processed segmentations maps from hcp subjects, etc.)

Same image, directly called afterwards, but it works

~~~bash
samuel my_secret_folder_path_which_contains_too_much_information $ [DICOM_gab] /home/samuel/git/mrtrix3/release/bin/mrview 2.nii.gz -debug
mrview: [DEBUG] No config file found at "/etc/mrtrix.conf"
mrview: [DEBUG] No config file found at ".mrtrix.conf"
mrview: [INFO] GL renderer:  Quadro K2200/PCIe/SSE2
mrview: [INFO] GL version:   3.3.0 NVIDIA 367.48
mrview: [INFO] GL vendor:    NVIDIA Corporation
mrview: [DEBUG] loading font into OpenGL texture...
mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location = 0) in vec2 pos;
layout(location = 1) in vec2 font_pos;
uniform float scale_x;
uniform float scale_y;
out vec2 tex_coord;
void main () {
  gl_Position = vec4 (pos[0]*scale_x-1.0, pos[1]*scale_y-1.0, 0.0, 1.0);
  tex_coord = font_pos;
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
in vec2 tex_coord;
uniform sampler2D sampler;
uniform float red, green, blue;
out vec4 color;
void main () {
  color.ra = texture (sampler, tex_coord).rg;
  color.rgb = color.r * vec3 (red, green, blue);
}

mrview: [DEBUG] font loaded
mrview: [INFO] opening image "2.nii.gz"...
mrview: [DEBUG] sanitising image information...
mrview: [DEBUG] loading image "2.nii.gz"...
mrview: [INFO] mrview: uncompressing image "2.nii.gz"
mrview: [INFO] mrview: uncompressing image "2.nii.gz" [done]
mrview: [DEBUG] image "2.nii.gz" loaded
mrview: [DEBUG] image "2.nii.gz" initialised with strides = [ 57600 -1 -240 ], start = 57599, using indirect IO
mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location = 0) in vec3 vertpos;
layout(location = 1) in vec3 texpos;
uniform mat4 MVP;
out vec3 texcoord;
void main() {
  gl_Position =  MVP * vec4 (vertpos,1);
  texcoord = texpos;
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
uniform float offset;
uniform float scale;
uniform sampler3D tex;
in vec3 texcoord;
out vec4 color;
void main() {
  if (texcoord.s < 0.0 || texcoord.s > 1.0 ||
      texcoord.t < 0.0 || texcoord.t > 1.0 ||
      texcoord.p < 0.0 || texcoord.p > 1.0) discard;
  color = texture (tex, texcoord.stp);
  float amplitude = color.r;
  if (isnan(amplitude) || isinf(amplitude)) discard;
  amplitude = clamp ( scale * (amplitude - offset), 0.0, 1.0);
  color.rgb = vec3 (amplitude);
}

mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location=0) in vec2 pos;
void main () {
  gl_Position = vec4 (pos, 0.0, 1.0);
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
out vec4 color;
void main () {
  color = vec4 (0.5, 0.5, 0.0, 1.0);
}

mrview: [DEBUG] compiling OpenGL vertex shader:
#version 330 core
layout(location=0) in vec3 data;
uniform float scale_x, scale_y;
out float amplitude;
void main () {
  gl_Position = vec4 (data.x*scale_x-1.0, data.y*scale_y-1.0, 0.0, 1.0);
  amplitude = data.z;
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
in float amplitude;
out vec3 color;
uniform vec3 colourmap_colour;
void main () {
  color.rgb = vec3 (amplitude);
}

mrview: [DEBUG] compiling OpenGL fragment shader:
#version 330 core
out vec3 color;
void main () {
  color = vec3(1.0, 1.0, 0.0);
}

mrview: [DEBUG] image "2.nii.gz" unloaded
~~~

I can do this 4-5 times in a row, and it just seems to work on and off weirdly enough.

If anyone else has it working I'd like to know which version of the dependency if possible, it could just be that the default ubuntu build has a slight dependency build options/config problem, and as it is supported for 5 years, others might run into similar problems, who knows.

**Comment by jdtournier** [comment]
That's the problem:

> Checking for Eigen 3 library: 3.2.92

Ubuntu 16.04 ships with a beta version of Eigen 3.3 - not their stable release. I'm sure the problems we have are not related to whether it's stable or not, but with the use of CPU instructions that require stricter memory alignment. See [this thread](http://community.mrtrix.org/t/mrview-segmentationfault-core-dump-ubuntu-16-04/185/18?u=jdtournier) on the community (amongst others). There is also an [issue on GitHub](https://github.com/MRtrix3/mrtrix3/pull/578) about it, which we should probably reopen as it seems clear that the fixes I'd introduced aren't sufficient...

**Comment by samuelstjean** [comment]
Ah well, that is indeed probably it. I also just tried with eigen 3.3-rc1 (according to their own website eigen 3.3 is close to release), but it still segfaults unfortunately. Is is a large improvement though, instead of working 1/5, it segfaults 1/5. 

I hope they will get the proper non-beta release in time, for now I guess I'll just manually go back to 3.2. No idea they shipped non released software in a LTS, anyway, sorry for not checking the community thingy for duplicates issue. Since it's a dependency issue, I'd say this can be closed as I guess ubuntu will get the stable (and hopefully working) update in due time.

**Comment by jdtournier** [comment]
I'll close this and reopen the [original issue](https://github.com/MRtrix3/mrtrix3/pull/578), but just to be clear: the issue is _not_ that Eigen 3.3 is unstable (I'd be surprised if that was the case), but that the use of newer CPU instructions requires much greater attention to [memory alignment](http://eigen.tuxfamily.org/dox-devel/group__TopicUnalignedArrayAssert.html). We do need to fix the code to work with Eigen 3.3...
